kythe/rust/indexer/src/indexer/analyzers.rs

```diff
@@ -731,62 +797,103 @@ impl<'a, 'b> CrateAnalyzer<'a, 'b> {
         for reference in &analysis.refs {
             let mut reference_vname = template_vname.clone();
             let span = &reference.span;
+            let ref_id = reference.ref_id;
+
+            // Get the CrateId for the referenced crate
+            let krate_id_option = self.krate_ids.get(&ref_id.krate);
+            if krate_id_option.is_none() {
+                // This is a little bit of chicken and egg here. We could try
+                // and find the file vname first before the crate id, but then
+                // we wouldn't have a definition anchor to emit the diagnostic
+                // node to because we wouldn't have the crate id yet. The chance
+                // of a crate id not being in our HashMap is very low, so if this
+                // fails we just emit a diagnostic node to the first source file
+                // of the crate we are indexing
+                let file_vname = self.file_vnames.values().next().unwrap();
+                self.emitter.emit_diagnostic(
+                    file_vname,
+                    "Cross reference could not be generated",
+                    Some(&format!("Failed to generate cross reference for \"{:?}\" because the referenced crate could not be found", ref_id)),
+                    None
+                )?;
+            }
+            let krate_id = krate_id_option.unwrap();
 
             // Create VName for target of reference
             let mut target_vname = template_vname.clone();
-            target_vname
-                .set_signature(format!("{}_def_{}", krate_signature, reference.ref_id.index));
+            target_vname.set_signature(self.create_def_signature(krate_id, ref_id.index));
             target_vname.set_language("rust".to_string());
 
             // Create VName for the reference node
-            let file_vname = self.file_vnames.get(span.file_name.to_str().unwrap());
+            let file_name = span.file_name.to_str().unwrap();
+            let file_vname = self.file_vnames.get(file_name);
```

Comment:
It looks like every use of `self.file_vnames.get()` is based on a (rust analysis) `span`. Should we have  a method
`file_vname_from_span(span)` ?